domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,137 +196,364 @@ class TopicListController @Inject constructor(
+        listOfTopicIds.add(ADDITION_AND_SUBTRACTION_TOPIC_ID)
+        listOfTopicIds.add(MULTIPLICATION_TOPIC_ID)
+        listOfTopicIds.add(DIVISION_TOPIC_ID)
+      }
+      RATIOS_TOPIC_ID -> {
+        listOfTopicIds.add(ADDITION_AND_SUBTRACTION_TOPIC_ID)
+        listOfTopicIds.add(MULTIPLICATION_TOPIC_ID)
+        listOfTopicIds.add(DIVISION_TOPIC_ID)
+      }
+      ADDITION_AND_SUBTRACTION_TOPIC_ID -> {
+        listOfTopicIds.add(PLACE_VALUE_TOPIC_ID)
+      }
+      MULTIPLICATION_TOPIC_ID -> {
+        listOfTopicIds.add(ADDITION_AND_SUBTRACTION_TOPIC_ID)
+      }
+      DIVISION_TOPIC_ID -> {
+        listOfTopicIds.add(MULTIPLICATION_TOPIC_ID)
+      }
+      EXPRESSION_AND_EQUATION_TOPIC_ID -> {
+        listOfTopicIds.add(ADDITION_AND_SUBTRACTION_TOPIC_ID)
+        listOfTopicIds.add(MULTIPLICATION_TOPIC_ID)
+        listOfTopicIds.add(DIVISION_TOPIC_ID)
+      }
+      DECIMALS_TOPIC_ID -> {
+        listOfTopicIds.add(ADDITION_AND_SUBTRACTION_TOPIC_ID)
+        listOfTopicIds.add(MULTIPLICATION_TOPIC_ID)
+        listOfTopicIds.add(DIVISION_TOPIC_ID)
+      }
+    }
+    return listOfTopicIds
+  }
+
+  private fun computeSuggestedStories(
+    topicProgressList: List<TopicProgress>
+  ): List<PromotedStory> {
+    val recommendedStories = mutableListOf<PromotedStory>()
     val topicIdJsonArray = jsonAssetRetriever
       .loadJsonFromAsset("topics.json")!!
       .getJSONArray("topic_id_list")
-    for (i in 0 until topicIdJsonArray.length()) {
-      createRecommendedStoryFromAssets(topicIdJsonArray[i].toString())?.let {
-        recommendedStories.add(it)
+    val topicIdList =
+      (0 until topicIdJsonArray.length()).map { topicIdJsonArray[it].toString() }
+
+    // Suggest prerequisite topic user needs to learn after completing any of the topics.
+    for (i in 0 until topicIdList.size) {
+      val topicInProgressFound = topicProgressList.any { it.topicId == topicIdJsonArray[i] }
+      if (!topicInProgressFound) {
+        val dependentListOfTopics =
+          retrieveTopicDependencies(topicIdList[i])
+        for (j in dependentListOfTopics.indices) {
+          topicProgressList.mapIndexed { index, topicProgress ->
+            if (topicProgress.topicId == dependentListOfTopics[j]) {
+              val topic = topicController.retrieveTopic(topicProgressList[index].topicId)
+              if (checkIfAtLeastOneStoryIsCompleted(topicProgressList[index], topic)) {
+                val recommendedTopicAlreadyExist =
+                  recommendedStories.any { it.topicId == topicIdList[i] }
+                if (!recommendedTopicAlreadyExist) {
+                  val recommendedStoriesIdFromAssets =
+                    createRecommendedStoryFromAssets(topicIdList[i])
+                  if (recommendedStoriesIdFromAssets != null) {
+                    recommendedStories.add(recommendedStoriesIdFromAssets)
+                  }
+                }
+              }
+            }
+          }
+        }
```

Comment:
```suggestion
    // The list of started or completed topic IDs.
    val startedTopicIds = topicProgressList.map(TopicProgress::getTopicId)
    // All topics that could potentially be recommended.
    val topicIdList =
      (0 until topicIdJsonArray.length()).map { topicIdJsonArray[it].toString() }
    // The list of topic IDs that qualify for being recommended.
    val unstartedTopicIdList = topicIdList.filterNot { startedTopicIds.contains(it) }

    // Suggest prerequisite topic user needs to learn after completing any of the topics.
    for (topicId in unstartedTopicIdList) {
      // All of the topic's prerequisites must be completed before it can be suggested.
      val dependentTopicIds = retrieveTopicDependencies(topicId)
      if (startedTopicIds.containsAll(dependentTopicIds)) {
        createRecommendedStoryFromAssets(topicIdList[i])?.let {
          recommendedStories.add(it)
        }
      }
    }
```

It seems like this can be simplified. Would the above work?

Reply:
Per our discussion I just realized that my code example doesn't properly filter out earlier topics if later ones are completed.

Reply:
Updated the suggestion--the solution is more complex than I originally realized so I think there might have been some issues in the current implementation checked into the branch. Note that the above probably needs to be cleaned up slightly & might not work out-of-the-box/need some debugging.

Reply:
Removed the classes--those were part of a first attempt by me and aren't actually relevant.